### PR TITLE
Fix Tokenizer::int64() parsing of "0" when guessing base

### PR DESCRIPTION
**Pull request by rousskov**
Known bug victims in current code were tcp_outgoing_mark,
mark_client_packet, clientside_mark, and mark_client_connection
directives as well as client_connection_mark and (deprecated)
clientside_mark ACLs if they were configured to match a zero mark using
"0" or "0/..." syntax:

    ERROR: configuration failure: NfMarkConfig: invalid value '0/10'...
    exception location: NfMarkConfig.cc(23) getNfmark

Probably broken since 2014 commit 01f2137d.

### CODE REVIEW

File: src/parser/Tokenizer.cc [hunk 1]
```diff
@@ -264,7 +264,6 @@ Parser::Tokenizer::int64(int64_t & result, int base, bool allowSign, const SBuf:
     if (base == 0) {
         if ( *s == '0') {
             base = 8;
-            ++s;
```
Comment (rousskov) [inline src/parser/Tokenizer.cc:267]: We do not need to advance input here because our integer value computing math below should handle any leading zeros correctly. Advancing input leads to the bug this PR is fixing.

An alternative fix is to treat "0" and similar inputs specially. The fix proposed by this PR feels like a slightly lesser evil to me.

P.S. This comment is not requesting any changes.

Comment (yadij) [reply src/parser/Tokenizer.cc:267]: IIRC, this was for correct auto detection of octal "0" and "00" cases.

**Review by yadij — CHANGES_REQUESTED** [review]
Just some missing tests.

**Review by rousskov — COMMENTED** [review]
@yadij, I hope all of your concerns were addressed as of branch commit f9cc5eb1. Please re-review.

File: src/tests/testTokenizer.cc [hunk 2]
```diff
@@ -248,6 +248,285 @@ TestTokenizer::testTokenizerInt64()
...
+        CPPUNIT_ASSERT(!t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(SBuf("0x"), t.buf());
+    }
+
+    // invalid (when forcing hex base) input matching hex base
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0x"));
+        CPPUNIT_ASSERT(!t.int64(rv, 16));
+        CPPUNIT_ASSERT_EQUAL(SBuf("0x"), t.buf());
```
Comment (rousskov) [inline src/tests/testTokenizer.cc:511]: The situation is worse than I thought. The bug discovered by Eduard is not the only previously unknown difference with strtol().

While our int64() rejects `0x` when being forced to parse hex integers (see the test case this change request is attached to), strtol() successfully parses that as ordinary zero digit (followed by "x" trailing garbage).

strtol() does not require `0x` base indicator to be present when forced to parse hex. That makes sense, of course: Our own int64() also successfully parses indicator-free "deadbeefrow" sequence (the last test case in this source file), for similar "indicator is optional when the base is known" reasons.

The outcomes differ in the `0x` input case because our int64() insists on interpreting existing `0x` prefix as a base indicator while strtol() is happy to interpret it as a case where there is no base indicator _if_ that results in a successful parsing outcome (i.e., valid hex digit "0" followed by a garbage character "x"). 

The same applies to `0x` input when int64() is told to guess the base with base=0 argument. We reject that input (see the test case immediately above the one this change request is attached to). strtol() accepts `0x` as zero followed by garbage.

In summary, unlike our Tokenizer::int64(), strtol() does not fail if its input starts with zero, _regardless_ of the valid base parameter value. Similar statements can be made for other leading digits (and non-conflicting valid base values).

Those secret differences is another set of int64() problems that we should address, but it would require serious work. Should we do that in this PR? If yes, do you agree that int64() should mimic strtol() behavior in all cases where input starts with a non-space character? CC: @kinkie

**Comment by rousskov** [comment]
With @eduard-bagdasaryan's help, I added four directives to the two ACLs declared as known to be affected in this PR description: tcp_outgoing_mark, mark_client_packet, clientside_mark, and mark_client_connection. Their parser is parse_acl_nfmark() which calls Ip::NfMarkConfig::Parse() which ends up calling getNfmark() that call int64() with zero base.

File: src/tests/testTokenizer.cc [hunk 3]
```diff
@@ -264,6 +543,36 @@ TestTokenizer::testTokenizerInt64()
...
+    // zero corner case: "negative" zero
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("-0"));
+        const int64_t benchmark = -0;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT_EQUAL(SBuf(""), t.buf());
+    }
+
```
Comment (eduard-bagdasaryan) [inline src/tests/testTokenizer.cc:575]: Don't we need also to test overflow with a zero-leading huge value?

Comment (rousskov) [reply src/tests/testTokenizer.cc:575]: Yes, better overflow tests would be useful, but they are outside the current PR scope IMO. Writing a good black-box leading-zero overflow test is not trivial. Adding a white-box leading-zero overflow test is not needed -- int64() loop is currently not counting the number of digits in any way and, hence, does not care if there are leading zeros AFAICT. Let's not add those tests (in this PR).

P.S. I added (useful) `+0` and `-0` tests because adding them was _trivial_. They are arguably in this PR scope.

File: src/tests/testTokenizer.cc (hunk 2 above)
Comment (rousskov) [reply src/tests/testTokenizer.cc:511]: > > While our int64() rejects 0x when being forced to parse hex integers, strtol() successfully parses

> It is a question whether '0x' should be parsed as a valid (0) value or rejected

I wish that were true, but the actual question is rather different! You are thinking of parsing isolated integers (for the lack of a better term). In that context, there is no doubt that "0x" is _not_ a valid integer! However, Tokenizer::int64() and strtol() do not parse isolated integers. They parse input stream prefixes. When parsing an input stream, the question is which leading portion (if any) of the given input stream is an integer. For two-character "0x" input, "0" is that integer portion (and "x" is the remaining input).

Even after int64() is fully fixed, in most (perhaps even all currently existing) int64() call contexts, the entire "0x..." input will be rejected as invalid _after_ extracting/parsing a valid integer "0" because the remaining "x..." input will not satisfy input grammar, but that caveat does _not_ affect correct int64() functionality. The method is buggy.

Comment (rousskov) [reply src/tests/testTokenizer.cc:511]: > Those secret differences is another set of int64() problems that we should address. Documenting these differences is an easy way out (at least for now). Changing how int64() works would require serious work. Should we do any of that in this PR?

After giving it more thought over the weekend, I am now convinced that we should _not_ add more int64() fixes to this PR. Instead, we should address these remaining problems in followup PR(s): Unlike the "0" parsing bug fixed by this PR, the remaining problems are not trivial to fix, they do not affect known/common valid input cases, they do not block other existing PRs, and they may warrant a rather different approach to unit testing (which requires even more development work). They deserve dedicated PR(s).

@yadij, please disregard those other int64() problems and clear this PR, unblocking #1840 that depends on it.

Comment (eduard-bagdasaryan) [reply src/tests/testTokenizer.cc:511]: >  When parsing an input stream, the question is which leading portion (if any) of the given input stream is an integer.

Ok, it should be then int64() caller's decision whether, say, '0x' is an invalid integer (the end of input), or a maybe a valid one (more input is expected).

**Review by eduard-bagdasaryan — APPROVED** [review]
LGTM

File: src/tests/testTokenizer.cc (hunk 2 above)
Comment (rousskov) [reply src/tests/testTokenizer.cc:511]: Yes, those two are both valid caller use cases.

File: src/tests/testTokenizer.cc [hunk 4]
```diff
@@ -248,6 +248,68 @@ TestTokenizer::testTokenizerInt64()
...
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("1"));
+        const int64_t benchmark = 1;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT(t.buf().isEmpty());
+    }
+
+    // autodetect hex base in minimal valid input
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/tests/testTokenizer.cc]: resolved by 7928426f57237418e1f55120d06c0cefeb75ceac

File: src/tests/testTokenizer.cc (hunk 2 above)
Comment (yadij) [reply src/tests/testTokenizer.cc:511]: Please do not fixate on bug-compatibility with `strtol()` - the `Tokenizer` is for protocol parsing. As such the limits and tolerances come from those protocol specifications, not `strtol()` behaviour.

IME, network protocols typically expect "0x" to be detected as a broken/invalid syntax. Which our `Tokenizer` needs to support detection of.  So "0x" input should indeed be detected as incomplete for base-16 or auto-detection. All other (provided non-16 bases) should detect it as `0` with trailing character(s).

File: src/tests/testTokenizer.cc [hunk 5]
```diff
@@ -248,6 +248,68 @@ TestTokenizer::testTokenizerInt64()
...
 
+    // autodetect octal base in minimal valid input
+    {
+        int64_t rv;
+        Parser::Tokenizer t(SBuf("0"));
+        const int64_t benchmark = 0;
+        CPPUNIT_ASSERT(t.int64(rv));
+        CPPUNIT_ASSERT_EQUAL(benchmark,rv);
+        CPPUNIT_ASSERT(t.buf().isEmpty());
+    }
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/tests/testTokenizer.cc]: I tried to add an equivalent case earlier, when enumerating all possible interesting short cases. Look for "auto-detected base; input: base, no post-base digits, unlimited". That case had wrong base (8 instead of zero). Fixed in c409c333.

File: src/tests/testTokenizer.cc (hunk 2 above)
Comment (rousskov) [reply src/tests/testTokenizer.cc:511]: > Please do not fixate on bug-compatibility with `strtol()`

Please do not review people.

>  the `Tokenizer` is for protocol parsing. 

Tokenizer use cases are not limited to "protocol parsing". The class is already used for parsing squid.conf input, for example.

> IME, network protocols typically expect "0x" to be detected as a broken/invalid syntax. Which our `Tokenizer` needs to support detection of. So "0x" input should indeed be detected as incomplete for base-16 or auto-detection. All other (provided non-16 bases) should detect it as `0` with trailing character(s).

There are "0x is invalid input" cases you mention, but there may be other use cases as well. IMO, we should address the needs of use cases we already have (while staying as predictable/general/etc. as feasible). Doing that well is already difficult. We have [started to discuss](https://github.com/squid-cache/squid/pull/1588#discussion_r1397356692) related improvements, and I still plan to propose how to handle those needs better than we do today. All that is outside _this_ bug-fixing PR scope though, so I am not going to argue about some quoted statements that I disagree with.

**Comment by rousskov** [comment]
@yadij, please unblock this head-of-queue PR. It has no known blocking problems and has been waiting for your review for more than a week.

File: src/tests/testTokenizer.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/tests/testTokenizer.cc]: Thank you.

**Review by yadij — APPROVED** [review]

**Review by yadij — APPROVED** [review]

**Comment by yadij** [comment]
FYI, resync with master to kick a retry of the failed tests on [c409c33](https://github.com/squid-cache/squid/pull/1842/commits/c409c3334d8254d1e8e22fc28162b14d1435179f)
